module/zfs/dmu_send.c

```diff
@@ -2664,20 +2685,53 @@ dmu_send(const char *tosnap, const char *fromsnap, boolean_t embedok,
 	dspp.resumeobj = resumeobj;
 	dspp.resumeoff = resumeoff;
 	dspp.rawok = rawok;
+	dspp.savedok = savedok;
 
 	if (fromsnap != NULL && strpbrk(fromsnap, "@#") == NULL)
 		return (SET_ERROR(EINVAL));
 
 	err = dsl_pool_hold(tosnap, FTAG, &dspp.dp);
 	if (err != 0)
 		return (err);
+
 	if (strchr(tosnap, '@') == NULL && spa_writeable(dspp.dp->dp_spa)) {
 		/*
 		 * We are sending a filesystem or volume.  Ensure
 		 * that it doesn't change by owning the dataset.
 		 */
-		err = dsl_dataset_own(dspp.dp, tosnap, dsflags, FTAG,
-		    &dspp.to_ds);
+		if (savedok) {
+			char name[ZFS_MAX_DATASET_NAME_LEN + 6];
+
+			(void) snprintf(name, sizeof (name),
+			    "%s/%%recv", tosnap);
+
+			err = dsl_dataset_own_force(dspp.dp, name, dsflags,
+			    FTAG, &dspp.to_ds);
+			if (err == ENOENT) {
+				err = dsl_dataset_own_force(dspp.dp, tosnap,
+				    dsflags, FTAG, &dspp.to_ds);
+			}
+
+			if (err == 0) {
```

Comment:
if `err = ENOENT` above, we still take this branch. Is that correct? Because we don't execute the code in this branch after line 2732

Is `DS_FIELD_RESUME_TOGUID` even defined for the `tosnap`'s (as opposed to the `name` dataset)?

Reply:
@tcaputi can probably speak to this better, but my understanding is that when you do `zfs send --saved pool/fs`, there are 3 cases:
1. there is no saved state, give an error
2. there was an incremental receive on an existing fs in progress, therefore we have saved state on `pool/fs/%recv`
3. there was receive that created a new fs in progress (typically a full receive, but could be an incremental which creates a clone), therefore we have saved state on `pool/fs` (which has no snapshots or children)

I think this code is handling these correctly, but would appreciate your verification.